File: swift/lifegame.swift

import Foundation

class Lifegame {
    let height: Int
    let width: Int
    var field: Array<Array<Int>>

    init(height: Int, width: Int) {
        self.height = height
        self.width  = width
        self.field = Array(repeating: Array(repeating: 0, count: width), count: height)
        for y in 0 ..< height {
            for x in 0 ..< width {
                self.field[y][x] = self.random(max: 2)
            }
        }
    }

    func random(max maxNumber: Int) -> Int {
        return Int(arc4random_uniform(UInt32(maxNumber)))
    }

    func loop() {
        while true {
            clearScreen()
            dumpField()
            usleep(100000)
            self.field = evolve()
        }
    }

    func evolve() -> Array<Array<Int>> {
        var newField: Array<Array<Int>> =
            Array(repeating: Array(repeating: 0, count: width), count: height)
        for y in 0 ..< height {
            for x in 0 ..< width {
                let aliveNeighbours = countAliveNeighbours(y: y, x: x)
                if aliveNeighbours <= 1 { newField[y][x] = 0 }
                if aliveNeighbours == 2 { newField[y][x] = field[y][x] }
                if aliveNeighbours == 3 { newField[y][x] = 1 }
                if aliveNeighbours >= 4 { newField[y][x] = 0 }
            }
        }
        return newField
    }

    func countAliveNeighbours(y: Int, x: Int) -> Int {
        var count = 0
        for yi in -1 ... 1 {
            for xi in -1 ... 1 {
                if yi == 0 && xi == 0 { continue }
                if field[mod(y + yi, divBy: height)][mod(x + xi, divBy: width)] == 1 {
                    count += 1
                }
            }
        }
        return count
    }

    func mod(_ a: Int, divBy b: Int) -> Int {
        if a > 0 { return a % b }
        if a < 0 { return (a + b) % b }
        return 0
    }

    func dumpField() {
        for y in 0 ..< height {
            for x in 0 ..< width {
                print((field[y][x] == 0) ? " " : "o", terminator: "")
            }
            print() // puts new line
        }
    }

    func clearScreen() {
        print("\u{001B}[;H\u{001B}[2J")
    }
}

var lifegame = Lifegame(height: 20, width: 40)
lifegame.loop()

File: typescript/lifegame.ts

class Lifegame {
  field: number[][];
  constructor(private height: number, private width: number) {
    this.field = this.initField(height, width);
  }

  initField(height: number, width: number): number[][] {
    let field: number[][] = [];
    for (let y = 0; y < height; y++) {
      for (let x = 0; x < width; x++) {
        if (!field[y]) field[y] = [];
        field[y][x] = this.getRandomInt(1);
      }
    }
    return field;
  }

  getRandomInt(max: number): number {
    return Math.floor(Math.random() * (max + 1));
  }

  loop() {
    this.clearScreen();
    this.dumpField();
    this.field = this.evolve();
    setTimeout(this.loop.bind(this), 100); // 100ms
  }

  evolve(): number[][] {
    let field   : number[][] = this.field;
    let newField: number[][] = [];
    for (let y = 0; y < this.height; y++) {
      for (let x = 0; x < this.width; x++) {
        if (!newField[y]) newField[y] = [];
        switch (this.countAliveNeighbours(field, y, x)) {
          case 2:  newField[y][x] = field[y][x]; break;
          case 3:  newField[y][x] = 1;           break;
          default: newField[y][x] = 0;           break;
        }
      }
    }
    return newField;
  }

  countAliveNeighbours(field, y, x) {
    let count = 0;
    for (let yi = -1; yi <= 1; yi++) {
      for (let xi = -1; xi <= 1; xi++) {
        if (yi === 0 && xi === 0) continue;
        if (field[this.mod(y + yi, this.height)][this.mod(x + xi, this.width)] === 1) {
          count++;
        }
      }
    }
    return count;
  }

  mod(a: number, b: number): number {
    return (a + b) % b;
  }

  dumpField() {
    let field = this.field;
    let lines = field.map(function (line) {
      return line.map(function (cell) {
        return (cell === 0) ? ' ' : 'o';
      }).join('');
    });
    console.log(lines.join("\n"));
  }

  clearScreen() {
    // console.log('\033[;H\033[2J');
    console.log('\x1B[2J\x1B[0f\u001b[0;0H');
  }
}


let lifegame = new Lifegame(20, 40)
lifegame.loop()
